Introduce new Event object

WPprodigy opened the pull request:
## Description

### The Event object

Introduces a new `Event` object that helps standardize and simplify how we interact w/ events across the plugin, in a repeatable and testable way. Some examples:

```
// Making new event
$event = new Event();
$event->set_action( 'example_action' );
$event->set_timestamp( time() );
$event->save();

// Updating the event
$event = new Event( $id );
$event->set_timestamp( time() + 500 );
$event->save();

// Getting a specific event:
Event::get( [
  'timestamp' => time(),
  'action'    => 'example_action',
  'args'      => ['some args'],
] );
```

I usually wouldn't opt for the full getter/setter structure, but the object will help keep a strict contract w/ event interaction. Anything that happens at the single event-level is contained here, `$event->reschedule()` for example helps encapsulate the related logic.

Also of note is the general plan for the plugin to pass around these Event objects or an ID rather than random bits of args. So the orchestrator for example will just get the event ID and run via CLI w/ that. Helping side-step a range of problems we've run into w/ the current setup. So this may be a bit chonky of a PR initially, but a lot of existing code will be able to be removed as a result.

### The Events (plural) object
If something is happens to multiple events, then those will be going into the `Events` class (in a separate PR) where the interactions between multiple Event objects take place. Examples of the API that's going there to help give a full picture for the vision started in this PR:

```
// Query for multiple events.
Events::query( array $args ): array

// Format an array of Event objects the way WP expects the data
Events::format_events_for_wp( $events ): array
```


### The data store

Also going in a separate PR are the methods this object uses to interact with the actual database, but it may help to quickly see them here. The API exposed there is very simple, and notably intended to be private so all consumers go through Event() or Events().

```
_create_event( array $row_data ): int
_update_event( int $event_id, array $row_data ): bool
_get_event_raw( int $id ): ?object
_query_events_raw( array $args = [] ): array
```
_____

By itself, this PR does nothing. But in the other PRs, you will see the object being used for the new functionality. In the future existing plugin logic can be safely converted over in piece-mail.

## Testing

You'll actually want to checkout the `next/complete-changeset` PR to test the full picture together currently. Once checked out, can play with the new object. I recommend doing that locally while reviewing this new file rather than just via GH.

**Expect tests to fail in the individual PRs.** The final PR will be the passing one :)

## PR Process Notes
This is one of a few parts of changes that I have currently to ship together. To help keep things "small" and reviewable, I've created a branch called `next/pull-request-base` that these PRs will be opened against. As they are reviewed and approved, I'll merge into this branch. Everything going into `next/pull-request-base` will be approved via PR, I've set up a temporary branch protection on it.

Once all the pieces are in, `next/pull-request-base` will be rebased against master and can become the new "cron control next" branch for pre-prod testing, eventually to be merged into master for the production release.

pschoffer commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+		}
+
+		if ( $this->exists() ) {
+			$success = Events_Store::_update_event( $this->id, $row_data );
```
I guess this goes somewhere else. But doesn't `_function_name` mean kinda internal as in private? As we call this outside of Events_Store it is ot private and I would guess the method should ot be underscored then. But I dont know much abut php naming convetntions to be honest.

WPprodigy replied:
Yeah, the underscore is meant to showcase it as being preferable private (though unable to strictly enforce). Unfortunately the data store is a publicly accessible singleton atm though, so doing a "fake private" initially felt best until it's able to be made truly enclosed.

pschoffer replied:
Feels more like a subjective preference so sounds good to me as it is.

pschoffer reviewed (changes requested):
Left few nitpicky notes

WPprodigy commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+		// Event ran a bit delayed, adjust accordingly (example: a 12h interval event running 6h late will be scheduled for +6h from now).
+		// TODO: Maybe we can simplify here later and just always return `$now + $interval`?
+		$elapsed_time_since_due = $now - $this->timestamp;
+		$remaining_seconds_into_the_future = ( $interval - ( $elapsed_time_since_due % $interval ) );
```
The difference I believe is that the formula ensure it's set to "in the future" still. I'm not against changing it, is just what core does here: https://github.com/WordPress/WordPress/blob/92f0b0cb48a3de660655cc658fc739695a393096/wp-includes/cron.php#L443

pschoffer replied:
Hmm that's right! It does cover the real stale events better. Sounds good then.

WPprodigy commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+		}
+
+		// Pick out just the data that has changed.
+		$row_data = [];
```
Two main reasons: Seemed overkill to always update every column even if they don't change, and this also lowers the chance of one write contradicting another.

pschoffer replied:
I dont think mysql cares too much from a performance stand point if you update only few or more columns.

> lowers the chance of one write contradicting another
Could you explain this more? IF we are worried about parallel execution wouldn't this more likelly result in inconsistent data as some columns might come from one update and some from another?

WPprodigy replied:
I also like that a call to `$event->save()` will fail if you try to set a prop that was invalid (and thus nothing was marked as updated).

```
$event = new Event( $id );
$event->set_timestamp( -100 );
$result = $event->save();
```

The consumer just needs to check the save result, not all the others methods, to tell if the update was successful.

_____

> If we are worried about parallel execution wouldn't this more likely result in inconsistent data as some columns might come from one update and some from another?

Realistically the only things that will be updated under normal conditions is:

1) The `timestamp` & `schedule/interval` during `->reschedule()` after a job has run.
2) The `status` changes when events are unscheduled.
2) In the future, I want `status` updating more, such as `pending -> running -> pending -> running -> cancelled` for recurring events and `pending -> running -> complete/failed/cancelled` for singles.

The FE requests could be making schedule changes in parallel w/ the runner. My idea is to make it the runner's job to pull from primary DB `srtm` to help account for this. So before it runs it can know for sure if the job is still meant to run, and when it's done running something, before updating back to `pending`, it will check for sure from primary if the job has not been cancelled yet, ensuring it does not overwrite that.

As such, I'd prefer to not update the status in the row unless the consumer calls `set_status()` specifically, w/ the intention of changing the status to help limit the conflict chances w/ the FE.

WPprodigy commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+		if ( in_array( $status, Events_Store::ALLOWED_STATUSES, true ) ) {
+			$this->status = $status;
+		} else {
+			$this->status = Events_Store::STATUS_PENDING;
```
Whatcha mean?

pschoffer replied:
I mean if the status is NOT in allowed statuses, we discard what the caller send in and set PENDING. Should we instead throw an error and say something like: "The status `burn_in_hell` that you provided is not an allowed status."

WPprodigy replied:
I guess it's a matter of defaults. Right now if you don't set a status, a new event is created as `pending` automatically. So my thought process was to just default a failed `set_status()` attempt as pending, as it's the safe (and only) core WP status. 

The rest of the setters kind of follow a pattern where they protect the defaults, and just won't set if given missing or invalid data, helping ensure data integrity w/o the consumer needs to check the response of each set() (instead only need to confirm if the event was able to be saved with `->save()`, which will fail if it didn't get the required fields).

That said, the behavior in set_status() is a bit different since it makes a more active decision in the process. I'm thinking could just remove the active decision and just not update the status when given an invalid one. If it's a new event, `STATUS_PENDING` will be applied automatically. If it's an existing event, then whatever it was before the invalid set attempt will stay present. That does seem more sane.

WPprodigy commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+	private ?int $id;
+	private string $status;
+
+	// TODO: Maybe we don't need action_hashed going forward?
```
Yes and no. We still need action_hashed since the DB table expects it, and other existing functionality uses it. Eventually once the rest of the related plugin logic is consuming this object though, it could potentially be removed.

WPprodigy commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+	 * @param int|object $event Event to initialize.
+	 */
+	public function __construct( $event = null ) {
+		if ( is_int( $event ) && $event > 0 ) {
```
Not completely sure whats best, but I do I agree there is room for improvement. I mirrored it off of some WooCommerce data classes that work similar to this.

The main need is that I want to be able to hydrate the object by just passing in the row DB row to it, so we avoid any extra queries after fetching a bunch of rows at once.

pschoffer replied:
I think I used a wrong term. I think the function I am after for static is `pure`. As in you always get the same result for the same input (or something slightly broader maybe). Which in turn means no (non-pure) dependencies like calls to a database or API. 

If you would like to test `Event::get(123)` in a unit test you now need to have a database working.  And even worse that applies to all methods that wound call Event::get(123). If we would go with the registry idea, either supply the registry as an argument or make this non-static method on some class (maybe `EventStore`) that has the registry as a dependency already would be my preference.

WPprodigy commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+		return $this->interval;
+	}
+
+	private function populate_data( object $data ): void {
```
I like that.

WPprodigy commented on includes/class-event.php:
```diff
@@ -0,0 +1,399 @@
+
+	private function populate_data( object $data ): void {
+		if ( ! isset( $data->status, $data->action, $data->args, $data->timestamp ) ) {
+			if ( isset( $data->ID ) ) {
```
So how you are you thinking the object should be hydrated w/ data from a DB row?

pschoffer commented:
Just to summarize the notes from the chat about this PR (that I still remember):
* Having constructor without logic, and have loader static methods - one for load from db and one to populate/clone/hydrate from already pre-fetched db_row
* Setter validations should either throw errors or be moved to the save method where they would throw errors.
* Keeping track over what columns to update seems to me to have 50/50 pros vs cons:) So I would error on the side of less code. But both ways are OK I think.

WPprodigy commented:
Addressed, definitely feels better now. Thanks for the guidance.

1) Loading up an Event from the DB is now done with `Event::get_from_db_row( $db_row )`. Else to get one event, `Event::get( $id )` or `Event::get( $args )`. And in an upcoming PR, to get multiple events: `Events::query( $args )`.

2) The setters don't validation anymore (though some do still hold the purpose of data integrity / ensuring coupled things are done together). Instead validation for the props is done during `->save()`, keeping it so only one return result needs to be checked by consumers, and errors are caught instead of swallowed up.

3) I dropped the change-tracking for now in favor of simplicity (and caught a hidden bug in the process 🙂 ). Figure we can always add the "feature" back if we find it would be good for performance (like maybe when $args is a really large array).

WPprodigy commented:
Per our chat, separated out the functionality in `get()` to be more clear. So now the function signatures are much more indicative of what they do exactly:

```
Event {
  get( int $event_id ): ?Event
  find( array $query_args ): ?Event
  get_from_db_row( object $data ): ?Event
}
```

pschoffer reviewed (approved):
:rock:
